src/lucas_kanade.jl

```diff
@@ -1,259 +1,240 @@
+                        first_pyramid, grid, level,
+                    )
+                    if min_eigenvalue < algorithm.eigenvalue_threshold
+                        @inbounds status[did] = false
+                        break
+                    end
+                else
+                    grid, offsets = new_grid, new_offsets
+                end
 
-function prepare_linear_system(corresponding_point, A, Ix, Iy, grid, offsets, etp)
-    bx = 0.0
-    by = 0.0
-    P, Q = size(A)
-    # Evaluates sum(δI.*I_x) and sum(δI.*I_y), where δI = A .- B
-    @inbounds begin
-        for q = 1:Q
-           for p = 1:P
-                Apq = A[p,q]
-                r = corresponding_point[1] + offsets[1][p]
-                c = corresponding_point[2] + offsets[2][q]
-                Bpq = etp(r, c) # Subpixel precision
-                bx = bx + (Apq - Bpq)*Ix[p,q]
-                by = by + (Apq - Bpq)*Iy[p,q]
+                estimated_flow = compute_flow_vector(
+                    putative_correspondence,
+                    first_pyramid, interploated_layer, level,
+                    grid, offsets, G_inv,
+                )
+                pyramid_contribution += estimated_flow
+                # Check if tracked point is out of image bounds.
+                if !lies_in(level_resolution, point + pyramid_contribution)
+                    @inbounds status[did] = false
+                    break
+                end
+                # Epsilon termination criteria.
+                abs(estimated_flow[1]) < algorithm.ϵ &&
+                    abs(estimated_flow[2]) < algorithm.ϵ && break
+            end
+            @inbounds begin
+            # Check if flow is too big.
+            if status[did] && is_lost(displacement[did], window2x)
+                status[did] = false
+            end
+            if status[did]
+                displacement[did] = displacement[did] + pyramid_contribution
+                level != 1 && (displacement[did] *= 2.0)
+            end
             end
         end
     end
-    b = SVector{2}(bx, by)
-end
 
-function in_image(img::AbstractArray{T, 2}, point::SVector{2, U}, window::Int) where {T <: Gray, U <: Union{Int, Float64}}
-    first_axis = first(axes(img))
-    second_axis = last(axes(img))
-    if point[1] < first(first_axis) || point[2] < first(second_axis) || point[1] > last(first_axis) || point[2] > last(second_axis)
-        return false
-    else
-        return true
-    end
+    displacement, status
 end
 
-function is_lost(img::AbstractArray{T, 2}, point::SVector{2, U}, min_eigenvalue::Float64, eigenvalue_threshold::Float64, window::Int) where {T <: Gray, U <: Union{Int, Float64}}
-    if !(in_image(img, point, window))
-        return true
-    else
-        val = min_eigenvalue/window
+function compute_partial_derivatives(Iy, Ix)
+    Iyy = imfilter(Iy .* Iy, Kernel.gaussian(4))
+    Ixx = imfilter(Ix .* Ix, Kernel.gaussian(4))
+    Iyx = imfilter(Iy .* Ix, Kernel.gaussian(4))
```

Comment:
Here we can use `imfilter!` to reuse the allocated memory, and use [factored kernels](https://juliaimages.org/ImageFiltering.jl/dev/#Factored-kernels) for more efficient filtering.

Reply:
Thanks for the suggestion! I'm now reusing arrays and using explicitly factored kernels.
However, turns out, from the documentation `If the kernel is a two dimensional array, imfilter will attempt to factor it; if successful, it will use the separable algorithm.`. Thus I didn't notice much of the performance improvements, but there was a slight decrease in the number of allocations. Reusing array improved performance quite a bit though.